src/charm.py

```diff
@@ -252,6 +333,312 @@ def _on_nginx_route_broken(self, _: Any) -> None:
         """Handle the nginx-route-broken event."""
         self._update_ingress()
 
+    def _on_get_certificate_action(self, event: ActionEvent) -> None:
+        """Triggered when users run the `get-certificate` Juju action.
+
+        Args:
+            event: Juju event
+        """
+        hostname = event.params["hostname"]
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            event.fail("Certificates relation not created.")
+            return
+        tls_rel_data = tls_certificates_relation.data[self.app]
+        if tls_rel_data[f"certificate-{hostname}"]:
+            event.set_results(
+                {
+                    f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
+                    f"ca-{hostname}": tls_rel_data[f"ca-{hostname}"],
+                    f"chain-{hostname}": tls_rel_data[f"chain-{hostname}"],
+                }
+            )
+        else:
+            event.fail("Certificate not available")
+
+    def get_additional_hostnames(self) -> list:
+        """Get a list containing all ingress hostnames.
+
+        Returns:
+            A list containing service and additional hostnames
+        """
+        # The relation will always exist when this method is called
+        relation = self._get_relation()  # type: ignore[arg-type]
+        definition = self._get_definition_from_relation(relation)  # type: ignore[arg-type]
+        hostnames = [definition.service_hostname]
+        hostnames.extend(definition.additional_hostnames)
+        return hostnames
+
+    def _on_certificates_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the TLS Certificate relation created event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            event.defer()
+            return
+        hostnames = self.get_additional_hostnames()
+        for hostname in hostnames:
+            private_key_password = self._tls.generate_password().encode()
+            private_key = generate_private_key(password=private_key_password)
+            private_key_dict = {
+                "password": private_key_password.decode(),
+                "key": private_key.decode(),
+            }
+            if JujuVersion.from_environ().has_secrets:
```

Comment:
Shouldn't we manage the case when juju doesn't support secrets ?

Reply:
See above

Reply:
Closing this comment, let's continue the discussion on the other thread